From a152462eba4463e4ab9292e76fb1a3e3b23aed64 Mon Sep 17 00:00:00 2001
From: Jeroen Demeyer <jdemeyer@cage.ugent.be>
Date: Thu, 15 Feb 2018 11:06:32 +0100
Subject: [PATCH] In Python 3, an unbound method is just the function

---
 Cython/Utility/ModuleSetupCode.c |  2 +-
 tests/run/cyfunction.pyx         | 19 +++++++++++++++++++
 tests/run/pure_py.py             | 12 ++++++++++++
 3 files changed, 32 insertions(+), 1 deletion(-)
 tests patch don't apply cleanly on top of 0.27.3 dropping them

diff --git a/Cython/Utility/ModuleSetupCode.c b/Cython/Utility/ModuleSetupCode.c
index 1661f40be..2aeabd2fe 100644
--- a/Cython/Utility/ModuleSetupCode.c
+++ b/Cython/Utility/ModuleSetupCode.c
@@ -617,7 +617,7 @@ static CYTHON_INLINE void * PyThread_tss_get(Py_tss_t *key) {
 #endif
 
 #if PY_MAJOR_VERSION >= 3
-  #define __Pyx_PyMethod_New(func, self, klass) ((self) ? PyMethod_New(func, self) : PyInstanceMethod_New(func))
+  #define __Pyx_PyMethod_New(func, self, klass) ((self) ? PyMethod_New(func, self) : (Py_INCREF(func), func))
 #else
   #define __Pyx_PyMethod_New(func, self, klass) PyMethod_New(func, self, klass)
 #endif
